packages/commonwealth/test/integration/api/cosmosCache.spec.ts

```diff
@@ -0,0 +1,180 @@
+require('dotenv').config();
+import chai from 'chai';
+import 'chai/register-should';
+import chaiHttp from 'chai-http';
+chai.use(chaiHttp);
+const expect = chai.expect;
+
+import { RedisCache } from 'common-common/src/redisCache';
+import { RedisNamespaces } from 'common-common/src/types';
+import { cacheDecorator } from 'common-common/src/cacheDecorator';
+import app, { resetDatabase } from '../../../server-test';
+import { connectToRedis } from '../../util/redisUtils';
+import { sleep } from '../../util/delayUtils';
+
+function verifyNoCacheResponse(res) {
+  expect(res.body).to.not.be.null;
+  expect(res).to.have.status(200);
+  expect(res).to.not.have.header('X-Cache', 'HIT');
+}
+
+async function verifyCacheResponse(key, res, resEarlier) {
+  expect(res).to.have.status(200);
+  expect(res).to.have.header('X-Cache', 'HIT');
+  const valFromRedis = await cacheDecorator.checkCache(key);
+  expect(valFromRedis).to.not.be.null;
+  expect(JSON.parse(valFromRedis)).to.be.deep.equal(res.body);
+  expect(JSON.parse(valFromRedis)).to.be.deep.equal(resEarlier.body);
+}
+
+describe('Cosmos Cache', () => {
+  const redisCache: RedisCache = new RedisCache();
+  const route_namespace: RedisNamespaces = RedisNamespaces.Route_Response;
+
+  before(async () => {
+    await resetDatabase();
+    await connectToRedis(redisCache);
+    cacheDecorator.setCache(redisCache);
+  });
+
+  after(async () => {
+    await redisCache.deleteNamespaceKeys(route_namespace);
+    await redisCache.closeClient();
+  });
+
+  describe('Cosmos Cache', () => {
+    describe('cosmosAPI', () => {
+      async function makeRPCRequest(
+        body,
+        path = '/cosmosAPI/juno',
+        headers = {
+          'content-type': 'text/plain;charset=UTF-8',
+          'accept-language': 'en-US,en;q=0.9',
+        }
+      ) {
+        return chai.request(app).post(path).set(headers).send(body);
+      }
+
+      async function rpcTestBody(body, key) {
+        const res1 = await makeRPCRequest(body);
+        verifyNoCacheResponse(res1);
+        const res2 = await makeRPCRequest(body);
+        await verifyCacheResponse(key, res2, res1);
+      }
+
+      const rpcProposalsCacheExpectedTest = async (proposalStatus: string) => {
+        const pathParam = `{"path":"/cosmos.gov.v1beta1.Query/Proposals","data":"${proposalStatus}","prove":false}`;
+        const body = `{"jsonrpc":"2.0","id":382288611593,"method":"abci_query","params":${pathParam}}`;
+        const key = `/cosmosAPI/juno_{"path":"/cosmos.gov.v1beta1.Query/Proposals","data":"${proposalStatus}","prove":false}`;
```

Comment:
sorry again, should "id" be part of cache key or just "proposalStatus" ?

Reply:
I misspoke before. This is correct. Only using the pathParam for the cache key. That is the part that changes per request.

Reply:
We could narrow it down to proposalStatus, but it loses a lot of context without the path `/cosmos.gov.v1beta1.Query/Proposals`.

Reply:
renamed variable "params" - more correct name.